Lib/warnings.py

```diff
@@ -701,17 +707,48 @@ def extract():
 # If either if the compiled regexs are None, match anything.
 try:
     from _warnings import (filters, _defaultaction, _onceregistry,
-                           warn, warn_explicit, _filters_mutated)
+                           warn, warn_explicit,
+                           _filters_mutated_unlocked,
+                           _acquire_lock, _release_lock,
+    )
     defaultaction = _defaultaction
     onceregistry = _onceregistry
     _warnings_defaults = True
+
+    class _Lock:
+        def __enter__(self):
+            _acquire_lock()
+            return self
+
+        def __exit__(self, *args):
+            _release_lock()
+
+    _lock = _Lock()
+
+    def _filters_mutated():
+        # Even though this function is part of the public API, it's used
+        # by a fair amount of user code.
+        with _lock:
+            _filters_mutated_unlocked()
+
 except ImportError:
     filters = []
     defaultaction = "default"
     onceregistry = {}
 
+    import _thread
+
+    # Note that this is a non-reentrant lock, matching what's used by
+    # _acquire_lock() and _release_lock().  Care must be taken to
+    # not deadlock.
+    _lock = _thread.LockType()
+
     _filters_version = 1
 
+    def _filters_mutated_unlocked():
+        global _filters_version
+        _filters_version += 1
+
     def _filters_mutated():
         global _filters_version
```

Comment:
This should hold the lock and call `_filters_mutated_unlocked`, currently the lock is missing. 

Alternatively you can just move `_filters_mutated` from line 728 later to call `_filters_mutated_unlocked` with lock held out of the block for importing C impl.

Reply:
There is other code that uses the `_unlocked` suffix for the same purpose but I like your suggested name better.  I'll change it.  Saying "unlocked" could definitely be confusing.

Reply:
I think it actually doesn't matter if the lock is held there or not.  At least I would assume the `+=` binop would be atomic.  Maybe it's less confusing to hold the lock?  Third party code that mutates `filters` and then calls `_filters_mutated()` is not thread safe anyhow.

Reply:
>  At least I would assume the += binop would be atomic. Maybe it's less confusing to hold the lock? 

Unfortunately `+=` binop isn't atomic so it needs extra locking.  

> Third party code that mutates filters and then calls _filters_mutated() is not thread safe anyhow.

Yes that isn't thread safe anyhow.